solr/test-framework/src/java/org/apache/solr/util/CallerMatcher.java

```diff
@@ -0,0 +1,257 @@
+   *     first call to {@link #checkCaller()} ()} from any caller will match.
+   */
+  public CallerMatcher(Collection<String> callerExprs, Collection<String> excludeExprs) {
+    for (String callerExpr : callerExprs) {
+      String[] exprCount = callerExpr.split(":");
+      if (exprCount.length > 2) {
+        throw new RuntimeException("Invalid count in callerExpr: " + callerExpr);
+      }
+      String[] clazzMethod = exprCount[0].split("\\.");
+      if (clazzMethod.length > 2) {
+        throw new RuntimeException("Invalid method in callerExpr: " + callerExpr);
+      }
+      Set<String> methods = exactCallers.computeIfAbsent(clazzMethod[0], c -> new HashSet<>());
+      if (clazzMethod.length > 1) {
+        methods.add(clazzMethod[1]);
+      }
+      if (exprCount.length > 1) {
+        try {
+          int count = Integer.parseInt(exprCount[1]);
+          maxCounts.put(exprCount[0], count);
+          callCounts.put(exprCount[0], new AtomicInteger(0));
+        } catch (NumberFormatException e) {
+          throw new RuntimeException("Invalid count in callerExpr: " + callerExpr, e);
+        }
+      }
+    }
+    for (String excludeExpr : excludeExprs) {
+      String[] clazzMethod = excludeExpr.split("\\.");
+      if (clazzMethod.length > 2) {
+        throw new RuntimeException("Invalid method in excludeExpr: " + excludeExpr);
+      }
+      Set<String> methods = excludeCallers.computeIfAbsent(clazzMethod[0], c -> new HashSet<>());
+      if (clazzMethod.length > 1) {
+        methods.add(clazzMethod[1]);
+      }
+    }
+  }
+
+  /**
+   * Returns the set of caller expressions that were tripped (reached their count limit). This
+   * method can be called after {@link #checkCaller()} returns a matching expression to obtain all
+   * expressions that exceeded their count limits.
+   */
+  public Set<String> getTrippedBy() {
+    return Collections.unmodifiableSet(trippedBy);
+  }
+
+  /** Returns a map of matched caller expressions to their current call counts. */
+  public Map<String, Integer> getCallCounts() {
+    return callCounts.entrySet().stream()
+        .collect(
+            Collectors.toMap(
+                e ->
+                    e.getKey()
+                        + (maxCounts.containsKey(e.getKey())
+                            ? ":" + maxCounts.get(e.getKey())
+                            : ""),
+                e -> e.getValue().get()));
+  }
+
+  /**
+   * Returns the matching caller expression when its count limit was reached, or empty if no caller
+   * or no count limit was reached. Each invocation increases the call count of the matching caller,
+   * if any. It's up to the caller to decide whether to continue processing after this count limit
+   * is reached. The matching expression returned by this call will be also present in {@link
+   * #getTrippedBy()}.
+   */
+  public Optional<String> checkCaller() {
+    return stackWalker.walk(
```

Comment:
the cyclomatic complexity here is a bit wild.  Can you please break this up; maybe calling an extracted method or two that limits the indentation level?